### PR TITLE
feat(validations): add line number output skaffold to validation errors

### PR DESCRIPTION
**Pull request by aaron-prindle**
This PR adds line # information to `pkg/skaffold/schema/validation/validation.go`.  This helps users as the error messages are more actionable (know exactly which file/indices/text the error is related to) and serves as a way of integrating the location information into the LSP server.  This way the LSP server can surface validation errors in realtime on the erroring text.

### CODE REVIEW

File: pkg/skaffold/schema/validation/validation.go [hunk 1]
```diff
@@ -570,44 +651,61 @@ func validateLogPrefix(lc latestV1.LogsConfig) []error {
...
 	if config.Metadata.Name != "" {
-		id = fmt.Sprintf("in module %q", config.Metadata.Name)
-	} else {
-		id = fmt.Sprintf("in unnamed config at index %d", config.SourceIndex)
+		module = fmt.Sprintf(" in module %q", config.Metadata.Name)
 	}
+	// else {
+	// 	// TODO(aaron-prindle) what is config.SourceIndex? Seems it is 0 in my testing usually...
+	// 	id = fmt.Sprintf("in unnamed config at index %d", config.SourceIndex)
+	// }
```
Comment (MarlonGamez) [inline pkg/skaffold/schema/validation/validation.go:693]: should this be removed?

**Review by MarlonGamez — DISMISSED** [review]
This looks good 👍🏼 Are there any particular things we should test ? I just did some basic validations and seemed to work

**Comment by MarlonGamez** [comment]
I modified the `getting-started` example `skaffold.yaml` to look like this 
```yaml
apiVersion: skaffold/v2beta26
kind: Config
build:
  cluster:
    timeout: 20m
  artifacts:
  - image: skaffold-example
    docker:
      dockerfile: Dockerfile
deploy:
  kubectl:
    manifests:
      - k8s-*
```

When running, my output looks like this:
```
❯ skaf-dev dev -d gcr.io/marlongamez-main
invalid skaffold config: source: /Users/marlongamez/Projects/skaffold/examples/getting-started/skaffold.yaml,: only one element in set artifact can be set. got DockerArtifact and KanikoArtifact
source: /Users/marlongamez/Projects/skaffold/examples/getting-started/skaffold.yaml,: found a 'docker' artifact, which is incompatible with the 'cluster' builder:

image: skaffold-example
context: .
docker:
  dockerfile: Dockerfile
kaniko: {}

To use the 'docker' builder, remove the 'cluster' stanza from the 'build' section of your configuration. For information, see https://skaffold.dev/docs/pipeline-stages/builders/
source: /Users/marlongamez/Projects/skaffold/examples/getting-started/skaffold.yaml,: only one element in set artifact can be set. got DockerArtifact and KanikoArtifact
source: /Users/marlongamez/Projects/skaffold/examples/getting-started/skaffold.yaml,: found a 'docker' artifact, which is incompatible with the 'cluster' builder:

image: skaffold-example
context: .
docker:
  dockerfile: Dockerfile
kaniko: {}

To use the 'docker' builder, remove the 'cluster' stanza from the 'build' section of your configuration. For information, see https://skaffold.dev/docs/pipeline-stages/builders/
```

It seems like the location information isn't being set properly/retried properly

**Comment by aaron-prindle** [comment]
> I modified the `getting-started` example `skaffold.yaml` to look like this
> 
> ```yaml
> apiVersion: skaffold/v2beta26
> kind: Config
> build:
>   cluster:
>     timeout: 20m
>   artifacts:
>   - image: skaffold-example
>     docker:
>       dockerfile: Dockerfile
> deploy:
>   kubectl:
>     manifests:
>       - k8s-*
> ```
> 
> When running, my output looks like this:
> 
> ```
> ❯ skaf-dev dev -d gcr.io/marlongamez-main
> invalid skaffold config: source: /Users/marlongamez/Projects/skaffold/examples/getting-started/skaffold.yaml,: only one element in set artifact can be set. got DockerArtifact and KanikoArtifact
> source: /Users/marlongamez/Projects/skaffold/examples/getting-started/skaffold.yaml,: found a 'docker' artifact, which is incompatible with the 'cluster' builder:
> 
> image: skaffold-example
> context: .
> docker:
>   dockerfile: Dockerfile
> kaniko: {}
> 
> To use the 'docker' builder, remove the 'cluster' stanza from the 'build' section of your configuration. For information, see https://skaffold.dev/docs/pipeline-stages/builders/
> source: /Users/marlongamez/Projects/skaffold/examples/getting-started/skaffold.yaml,: only one element in set artifact can be set. got DockerArtifact and KanikoArtifact
> source: /Users/marlongamez/Projects/skaffold/examples/getting-started/skaffold.yaml,: found a 'docker' artifact, which is incompatible with the 'cluster' builder:
> 
> image: skaffold-example
> context: .
> docker:
>   dockerfile: Dockerfile
> kaniko: {}
> 
> To use the 'docker' builder, remove the 'cluster' stanza from the 'build' section of your configuration. For information, see https://skaffold.dev/docs/pipeline-stages/builders/
> ```
> 
> It seems like the location information isn't being set properly/retried properly

This is definitely confusing but this is actually working as intended.  The way that line #s are found from a schema object is by doing a pointer lookup.  During `visitStructs`, we iterate over interface{} values that we convert to reflect.Value objects.  The changes made to `visitStructs` attempt to see of the interface{}/reflect.Value is addressable via `v.CanAddr()`.  This code is here:
```
			if v.CanAddr() {
				cfgErr = ErrorWithLocation{
					Error:    err,
					Location: cfg.YAMLInfos.Locate(v.Addr().Pointer()),
				}
			} else {
				log.Entry(context.TODO()).Debugf("unexpected issue - unable to get pointer to struct in visitStruct")
				cfgErr = ErrorWithLocation{
					Error:    err,
					Location: configlocations.MissingLocation(),
				}
```

If you run this with `skaffold dev -v debug` you will see the line:
`unexpected issue - unable to get pointer to struct in visitStruct`

TLDR: the line #s aren't shown in this specific case (this is only possible in certain cases for this single rule `visitStruct`) as it is a primitive type that has the error which is not addressable.  There is likely a way to get this to work by having visitStruct keep a parent ref but I would rather not do that in this PR.  I did not have plans on making this change atm, I will see if it is simple.  If not I might take this a TODO.  Thanks for investigating this though, this is important to resolve for 100% coverage for line #s w/ validation

File: pkg/skaffold/schema/validation/validation.go (hunk 1 above)
Comment (aaron-prindle) [reply pkg/skaffold/schema/validation/validation.go:693]: Yes, changed this to respect how it was before and comments are removed

**Comment by aaron-prindle** [comment]
In looking at this more there are two errors triggered in your sample, one is from `visitStructs` and the other is from `validateArtifactTypes`.  Both of these were not working as they were both mapping to `ArtifactType` object which have the following yamltag:
```
	// ArtifactType describes how to build an artifact.
	ArtifactType `yaml:",inline"`
```

structs with this tag are currently not exported in our pointer->line-# mapping as my original thought was that they wouldn't be flagged as erroring as they don't have a mappin in the yaml (their sub-elements do).  It seems though that we do look at this object for errors.  I fixed this for `validateArtifactTypes` by looking at the parent.  I still need to investigate `visitStructs`.  I hope to fix this but it is a somewhat minor issue as the actual skaffold validator catches this w/o line #s, it will only be the LSP that won't display this one rule.

**Comment by aaron-prindle** [comment]
Thanks for testing this Marlon, the issue that you identified is now fixed!! Thanks again for trying this functionality out, really appreciate it!

File: pkg/skaffold/schema/validation/validation.go [hunk 2]
```diff
@@ -570,44 +648,59 @@ func validateLogPrefix(lc latestV1.LogsConfig) []error {
...
 	}
+
 	for i := range errs {
-		errs[i] = errors.Wrapf(errs[i], "source: %s, %s", config.SourceFile, id)
+		if errs[i].Location == nil || errs[i].Location.StartLine == -1 {
+			errs[i].Error = errors.Wrapf(errs[i].Error, "source: %s,%s", config.SourceFile, id)
+			continue
+		}
+		errs[i].Error = errors.Wrapf(errs[i].Error, "source: %s,%s on line %d column %d",
+			config.SourceFile, id, errs[i].Location.StartLine, errs[i].Location.StartColumn)
```
Comment (MarlonGamez) [inline pkg/skaffold/schema/validation/validation.go:696]: put space after these commas to read better on the command line 
```suggestion
		if errs[i].Location == nil || errs[i].Location.StartLine == -1 {
			errs[i].Error = errors.Wrapf(errs[i].Error, "source: %s, %s", config.SourceFile, id)
			continue
		}
		errs[i].Error = errors.Wrapf(errs[i].Error, "source: %s, %s on line %d column %d",
			config.SourceFile, id, errs[i].Location.StartLine, errs[i].Location.StartColumn)
```

**Review by MarlonGamez — APPROVED** [review]
LGTM ✅ Have one nit that isn't a big deal but that slightly improves readability. Up to you if you'd like to fix

File: pkg/skaffold/schema/validation/validation.go (hunk 2 above)
Comment (aaron-prindle) [reply pkg/skaffold/schema/validation/validation.go]: good idea!  done
